### PR TITLE
only load changed i18n files when updated

### PR DESCRIPTION
**Pull request by wjordan**
This addition to the `fast_localization` initializer speeds up the amount of time it takes to reload changed i18n files by patching the `ActiveSupport::FileUpdateChecker` hook configured by the i18n system. Instead of calling `I18n.reload!` which reloads all of the i18n files in the entire `load_path`, this PR patches the existing hook to instead call `I18n.backend.load_translations(files)`, where `files` is the (much smaller) set of files that have been changed.

This should resolve some performance issues on levelbuilder where levels are constantly being updated many times per minute, and where the amount of time it takes to reload all translations is ~20-30 seconds.

### CODE REVIEW

File: dashboard/config/initializers/fast_localization.rb [hunk 1]
```diff
@@ -13,3 +13,30 @@
...
+  end
+end
+
+module UpdateCheckerExt
+  # Expose @block and @files variables.
+  attr_accessor :block, :files
+
+  # Pass list of updated files as a block parameter.
+  def execute
+    @last_watched = watched
```
Comment (ashercodeorg) [inline dashboard/config/initializers/fast_localization.rb:32]: @watched?

**Comment by wjordan** [comment]
merging now to avoid sync issues with levelbuilder branch. (will make any PR-feeback changes in future commits)

**Comment by joshlory** [comment]
Why are i18n files auto-reloaded on levelbuilder?  From what I can tell it's due to `cache_classes = false` which is inherited from [`environments/staging.rb`](https://github.com/code-dot-org/dashboard/commit/aa452f7c768c132988501a811af021413037af2b).

It makes sense that they need to be refreshed after they change, but I wonder if it would be better to manually reload the changed files.  Triggering auto-reload (and modifying how auto-reload works) seems like a roundabout way to achieve the same result.

IMHO there would be less magic if we manually reload the changed i18n data when it's updated.

File: dashboard/config/initializers/fast_localization.rb (hunk 1 above)
Comment (wjordan) [inline dashboard/config/initializers/fast_localization.rb:32]: this calls [`#watched`](https://github.com/rails/rails/blob/4-2-stable/activesupport/lib/active_support/file_update_checker.rb#L93-L99) in `ActiveSupport::FileUpdateChecker`.

the `execute` implementation overrides the [existing method](https://github.com/rails/rails/blob/4-2-stable/activesupport/lib/active_support/file_update_checker.rb#L70-L79) with the ability to pass a list of updated files as a block parameter.

**Comment by wjordan** [comment]
Thanks for the comments @joshlory!

Rails (ActiveSupport) i18n is hard-coded to always auto-reload all of the i18n files in its load path, not anything configured by `cache_classes` as far as I can tell (see [`i18n_railtie.rb`](https://github.com/rails/rails/blob/4-2-stable/activesupport/lib/active_support/i18n_railtie.rb)).

Definitely worth considering the alternative of manually-reloading changed i18n data when it's updated. Looks to me like that approach would require the following:
1. Disabling the hard-coded i18n auto-update functionality by removing the existing FileUpdateChecker using an initializer patch;
2. Adding hooks to the various places we update i18n `yml` files on the fly to manually reload affected files:
   1. [`Script#update_i18n`](https://github.com/code-dot-org/code-dot-org/blob/4dbb50a8216b8dda0cb586c3eec42ac62ac074b5/dashboard/app/models/script.rb#L583-L588)
   2. [`DSLDefined#rewrite_i18n_file`](https://github.com/code-dot-org/code-dot-org/blob/0c3271ac0ffe05de22f0f5a0260c392957ddd073/dashboard/app/models/dsl_defined.rb#L82-L90)
   3. [`seed:dsls`](https://github.com/code-dot-org/code-dot-org/blob/bf839e4235f6357b34fd01d0f2b2593deee65381/dashboard/lib/tasks/seed.rake#L75-L94)
   4. (Any future application logic where we create/modify i18n strings dynamically)

It's a tradeoff, but my own preference in this case is to stick with the approach of making a targeted patch to the underlying i18n implementation since it's a narrow optimization to the underlying implementation that could be eventually merged upstream. Since a patch is necessary in either case, I'd prefer to maintain a slightly longer initializer patch versus a shorter initializer patch combined with with extra code interleaved throughout our application logic.

I don't consider this PR (or either approach to this optimization) to be adding any significant amount of magic to our system, considering we are already depending on ActiveSupport i18n (and its associated implementation complexity) to begin with.

**Comment by wjordan** [comment]
Actually, my mistake- I looked into it a bit further and it does look like `cache_classes` affects the i18n updating mechanism. The `ActionDispatch::Reloader` middleware (which i18n hooks into to reload its changed files) is [only included in the middleware stack when `cache_classes` is false](https://github.com/rails/rails/blob/4-2-stable/railties/lib/rails/application/default_middleware_stack.rb#L40-L42).

This could make the manual-loading approach slightly simpler, I'll look into implementing that in a separate PR. Thanks for the suggestion!

**Comment by joshlory** [comment]
@wjordan I want to follow up on this.  I'm about to do some work with the Level Builder translation pipeline.  It would be helpful to reduce some of the magic in preparation for my change!

**Comment by wjordan** [comment]
My preference is to leave the current implementation in place, it solves the performance issue we experienced and hasn't raised any other specific issues (I addressed the 'magic' point above).

I briefly looked into the alternative discussed above and I don't think it will work as described. The default (`Simple`) i18n backend stores translations in a simple in-memory hash that isn't shared across worker processes. If the application updates an i18n `.yml` file and manually tries to reload the changed data, it will  only update its own in-memory hash, and the separate i18n backends across all other worker processes will still have stale data. You could work around this by switching i18n backends to storage that's shared across processes, but this would introduce further complexity to the fix.

If you'd like to work on an alternate fix as part of your related work feel free, I just don't think it's worth spending more time on myself.

**Pull request merged**